

```{r}
####CHUNK 1 : Libraries and output specification ####

library(dplyr)
library(tidyr)
library(stringr)
library(bio3d)
library(AlphaMissenseR)#Bioconductor necessary 
library(msa)           #Bioconductor necessary 
library(Biostrings)    #Bioconductor necessary 
library(tibble, include.only = "rownames_to_column")
library(ggplot2)


msa_output<-"../../Data/SH3_alignments/SH3_domains_25_aa_flanking.fasta"
structure_output_folder<-"../../Data/SH3_alignments/AF_db_structures/"
rmsd_output<-"../../Data/SH3_alignments/rmsd_data.csv"
sho1_pbs2_file<-"../../Data/AlphaFold_predictions/alphafold_output/WT.pdb"


```


```{r}

####CHUNK 2 : Import data ####

#From Interpro, list of yeast SH3 containing proteins, and where the SH3 domain is located in the sequence
interpro_SH3_location<-read.table("../../Data/SH3_alignments/protein-matching-PF00018.tsv", header=T, sep = "\t")%>%
  select(Accession, Length, Matches)

#From Uniprot, the matches of uniprot accessions into gene names
uniprot_id_to_gene_name<-read.table("../../Data/SH3_alignments/idmapping_2024_08_16.tsv", header=T)%>%
  dplyr::rename(Accession=From, Gene.symbol=To)


#From the alliancemine server, extract SH3 containing protein sequences, using the gene names
gene_name_seqs<-read.table("../../Data/SH3_alignments/alliancemine_seqs_id.tsv", header=T)%>%
  mutate(Gene.symbol=case_when(Gene.symbol == "" ~ Gene.secondaryIdentifier,
                                TRUE ~ Gene.symbol))%>% #YLL017W has no gene name so let's keep it as YLL017W for joining
  select(Gene.proteins.sequence.residues, Gene.symbol)%>%
  mutate(Gene.proteins.sequence.residues = str_remove_all(Gene.proteins.sequence.residues, "\\*"))





data<-left_join(interpro_SH3_location, uniprot_id_to_gene_name, by="Accession")%>%
  left_join(., gene_name_seqs, by="Gene.symbol")


data<-data%>%
  tidyr::separate_longer_delim(Matches, delim=",")%>%
  tidyr::separate_wider_delim(Matches, delim="..", names = c("SH3_start", "SH3_end"))


#For multiple SH3 per protein, identify _1 and _2
for(i in 1:(nrow(data)-1)){
  if(data[i, "Gene.symbol"]==data[i+1, "Gene.symbol"]){
    data[i, "Gene.symbol"]<-paste0(data[i, "Gene.symbol"], "_1")
    data[i+1, "Gene.symbol"]<-paste0(data[i+1, "Gene.symbol"], "_2")
  }
}


rm(gene_name_seqs, interpro_SH3_location, uniprot_id_to_gene_name, i)
```





```{r}
####CHUNK 3 : Identify flanking sequences####


#Add details about the beginning and end of the flanking sequences


data_detailed<-data%>%
   mutate(SH3_start=as.numeric(SH3_start),
         SH3_end=as.numeric(SH3_end))%>%
  mutate(flanking_before=case_when(SH3_start-25 < 1 ~ 
                                     str_sub(Gene.proteins.sequence.residues, start=1, end=SH3_start-1),
                                   SH3_start-25 > 1 ~ 
                                     str_sub(Gene.proteins.sequence.residues, start=SH3_start-25, end=SH3_start-1)))%>% #Make seq of SH3-25 unless too close to terminus
  mutate(flanking_start=case_when(SH3_start-25 < 1 ~ 1,
                                   SH3_start-25 > 1 ~ SH3_start-25))%>% #Start position 
  mutate(SH3_domain=str_sub(Gene.proteins.sequence.residues, start=SH3_start, end=SH3_end))%>% #SH3 seq
  mutate(flanking_after=str_sub(Gene.proteins.sequence.residues, start= SH3_end+1, end=SH3_end+25))%>% #Flanking sequence
  mutate(flanking_end=case_when(SH3_end+25 > str_length(Gene.proteins.sequence.residues) ~ 
                                    str_length(Gene.proteins.sequence.residues),
                                  SH3_end+25 <= str_length(Gene.proteins.sequence.residues) ~
                                    SH3_end+25))%>% #End of flanking
  mutate(flanking_before=str_to_lower(flanking_before),
         flanking_after=str_to_lower(flanking_after))%>% #Make flanking seqs lowercase
  unite("SH3_flanking_seq", flanking_before, SH3_domain, flanking_after, sep="")#Combine SH3+flanking


```

```{r}
#### CHUNK 4 : Produce FASTA for MSA####


#Place Sho1 at the top for easier reading when comparing MSAs
reordered_data_detailed<-rbind(filter(data_detailed, Gene.symbol == "SHO1"), 
                               filter(data_detailed, Gene.symbol != "SHO1"))

#Make a fasta of the Gene names with SH3 + flanking sequences
fasta_of_all<-c(rbind(reordered_data_detailed%>%
                 mutate(Gene.symbol=paste0(">", Gene.symbol))%>%
                 pull(Gene.symbol),
               reordered_data_detailed$SH3_flanking_seq))


write.table(fasta_of_all, file = msa_output, quote=F, row.names = F, col.names = F)

#Aligned using mafft 7.526 on WSL
#Command line: "/usr/bin/mafft"  --localpair --maxiterate 16  --inputorder "SH3_domains_25_aa_flanking.fasta" > "mafft_aligned_SH3_flanking.txt"

```


```{r}

####CHUNK 5 : Get (predicted) structures from AlphaFold DB ####

if(!dir.exists(structure_output_folder)){
  dir.create(structure_output_folder)}

get_trimmed_pdb_file<-function(SH3_domain, begin, end, uniprot_accession){
  if (SH3_domain != "SHO1"){
  pdb_url<-AlphaMissenseR::af_predictions(uniprot_accession)%>%
    pull(pdbUrl)  #Get address of AF DB prediction
  pdb_file<-BiocFileCache::bfcrpath(rnames = basename(pdb_url), fpath=pdb_url) #Save AF prediction
  
}else if (SH3_domain == "SHO1"){
 pdb_file<-sho1_pbs2_file
}
  full_pdb<-bio3d::read.pdb(pdb_file) #Import AF prediction
  
  sele<-bio3d::atom.select(full_pdb, resno=begin:end, chain="A") #Select only residues in the SH3 and flanking
  trimmed_pdb<-bio3d::trim(full_pdb, inds=sele)              #Trim pdb file to SH3 and flanking
  write.pdb(trimmed_pdb, paste0(structure_output_folder, SH3_domain, ".pdb")) #Save trimmed pdb
  
}





for (i in 1:nrow(data_detailed)){
  print(data_detailed[[i, "Gene.symbol"]])
  
  get_trimmed_pdb_file(SH3_domain = data_detailed[[i, "Gene.symbol"]],
                       begin = data_detailed[[i, "flanking_start"]],
                       end = data_detailed[[i, "flanking_end"]],
                       uniprot_accession = data_detailed[[i, "Accession"]] )
}


#Aligned using MUSTANG command line (from ../../Data/SH3_alignment/) :
#./bin/mustang-3.2.3 -f ./path_to_files.txt -F fasta -o SH3_flanking -s off




rm(i)
```



```{r}

#### CHUNK 6 : Get RMSD of 4 structures ####

#Get only Sho1 SH3 from predicted Sho1-Pbs2 interaction
sho1_pbs2_pdb<-read.pdb(sho1_pbs2_file)
sele<-atom.select.pdb(sho1_pbs2_pdb, chain = "A")
sho1_only_pdb<-bio3d::trim(sho1_pbs2_pdb, inds=sele)
sho1_sh3_start<-data_detailed%>%filter(Gene.symbol == "SHO1")%>%pull(SH3_start)
sho1_sh3_end<-data_detailed%>%filter(Gene.symbol == "SHO1")%>%pull(SH3_end)
sele<-atom.select.pdb(sho1_only_pdb,
                      resno = sho1_sh3_start:sho1_sh3_end)
sho1_sh3_pdb<-bio3d::trim(sho1_only_pdb, inds=sele)




#Get structures of SH3 domains from PDB
get_sh3_pdbs<-function(pdb_code, chain){
  current_pdb<-read.pdb(pdb_code)
  current_sele<-atom.select.pdb(current_pdb, chain = chain)
  trimmed_pdb<-bio3d::trim(current_pdb, inds=current_sele)
  return(trimmed_pdb)
}


abp1_pdb<-get_sh3_pdbs("2rpn", "A")
nbp2_pdb<-get_sh3_pdbs("2lcs", "A")
bem1_pdb<-get_sh3_pdbs("2rqw", "A")


#Align structures using msa package
pdbs<-pdbaln(list(sho1_sh3_pdb, abp1_pdb, nbp2_pdb, bem1_pdb), 
             exefile="msa")

#Calculate pairwise rmsd between structures
rmsd<-rmsd(pdbs, fit=T)%>%as.data.frame()

#Identify columns and rows
colnames(rmsd)<-c("Sho1", "Abp1", "Nbp2", "Bem1")
row.names(rmsd)<-c("Sho1", "Abp1", "Nbp2", "Bem1")

#Reshape dataframe 
long_rmsd<-rmsd%>%
  tibble::rownames_to_column(var="structure1")%>%
  pivot_longer(cols=c(Sho1, Abp1, Nbp2, Bem1), names_to = "structure2", values_to = "RMSD")

write.csv(long_rmsd, file = rmsd_output, row.names = F)
  
```



