```{r}

library(dplyr)
library(ggplot2)
library(tidyr)
library(ggpubr)
library(stringr)
library(ggExtra)
library(colorBlindness)
library(cowplot)
library(GGally)
library(RColorBrewer)
library(ggmsa)


motif_colour<-"skyblue2"
flanking_colour<-"#773344"

non_syn_colour<- motif_colour 
stop_colour<-"#D81B60"
syn_colour<-"#ffc107"
wt_colour<-"#004D40"
hog1_colour<-"#e69ba8"
sho1_colour<-"#028b5b"

theme_sup_figure<-function(textsize=12){
  theme(axis.text = element_text(size = textsize, colour="black"),
        axis.title = element_text(size=textsize+5),
        legend.title = element_text(size=textsize+2),
        legend.text = element_text(size=textsize),
        text = element_text(colour = "black"))
}
```

```{r}
####Supp fig 1 - Surrounding region heatmap####

#T-test to see if Hog1 score for mutants is sig different from wt
supp_fig1_t_test<-read.csv("../Data/DMS/results/codon_surrounding_region_DMS_scores.csv")%>%
  filter(condition == "MTX" & interactor == "Hog1" & sorbitol ==1)%>%
  filter(type !="WT" | position == 85)%>%                                        #Only 1 value for WT instead of 1 per position
  dplyr::mutate(identity=case_when(type %in% c("Non-Syn", "STOP") ~ paste(position, aa, sep="_"), #Separate Non-synonymous mutants,
                                   type %in% c("WT", "Syn") ~ "Syn+WT"))%>%          #WT and Syn mutants. the "reference" is WT+Syn pooled
  group_by(condition, interactor, sorbitol)%>%                                       #Group for T-test
  rstatix::pairwise_wilcox_test(log2FC ~ identity,                                   #Non-parametric t-test
                                ref.group = "Syn+WT",                                #Compare Non-syn mutants against Syn+WT   
                                p.adjust.method = "fdr",                             #Correction for multiple corrections
                                alternative ="two.sided")%>%                         #Two sided test                           
  rename(identity = group2)%>%                                                       #For joining later
  select(c(condition, interactor, sorbitol, identity, p.adj, p.adj.signif)) %>%      #Keep only relevant columns
  separate_wider_delim(., cols=identity, delim="_", names=c("position", "aa"))%>%
  mutate(position= as.numeric(position))%>%
  filter(p.adj.signif!="ns" & aa != "*")%>%
  select(position, aa)


supp_1_data<-read.csv("../Data/DMS/results/codon_surrounding_region_DMS_scores.csv")%>%
  filter(condition == "MTX" & sorbitol ==1 & interactor== "Sho1")%>%
          group_by(position, target, condition, sorbitol, interactor, aa, strain_background, type) %>%
          summarize(replicates_per_aa = n(),
                    median_log2FC=median(log2FC, na.rm=T)) %>%
          filter(replicates_per_aa >= 2) %>%
          unite("identifier", target, condition, sorbitol, interactor, strain_background, remove=F)
  

supp_fig1_wt_pbs2<-read.csv("../Data/DMS/reference/Pbs2_wt_seq.csv")%>%
  mutate(position=position+70)

supp_fig_1<-supp_1_data%>%
  ggplot(aes(x=as.factor(position), y=aa))+
  geom_tile(aes(fill=median_log2FC, colour=median_log2FC))+
  geom_text(data=supp_fig1_wt_pbs2, label="*")+
  geom_text(data=supp_fig1_t_test, label="H")+
  # geom_tile(data=supp_fig6_t_test, fill=NA, colour="black", linewidth=1)+
  scale_x_discrete(expand=c(0,0),breaks=seq(from=71, to=126, by=2) )+
  scale_y_discrete(expand = c(0,0))+
  scale_fill_gradient2(low="blue", mid="white", high="red")+
  scale_colour_gradient2(low="blue", mid="white", high="red", guide="none")+
  theme_sup_figure(textsize = 12)+
  theme(panel.border = element_rect(fill = NA, colour="black"),
        panel.background = element_rect(fill="black"),
        panel.grid.major = element_blank(),
        axis.text.x = element_text(size=13))+
    labs(x="Position in Pbs2",
         y="Residue",
         fill="Sho1-Pbs2\nInteraction score")


supp_fig_1

save_plot(filename = "../Data/images/SFigs/Supplementary_figure_1.png", base_width = 12,  plot = supp_fig_1, dpi=300)

rm(supp_1_data, supp_fig1_wt_pbs2, supp_fig_1, supp_fig1_t_test)
```


```{r, message=F, warning=F}
####Supp fig 2 - Surrouinding region replicates####
###Panel A
sup_2_data<-read.csv("../Data/DMS/results/codon_surrounding_region_DMS_scores.csv")%>%
  mutate(id=case_when(condition == "MTX" & sorbitol ==1 ~ "MTX + sorbitol",
                      condition == "MTX" & sorbitol ==0 ~ "MTX alone",
                      condition == "DMSO" & sorbitol ==1 ~ "DMSO + sorbitol",
                      condition == "DMSO" & sorbitol ==0 ~ "DMSO alone"))

#Sho1
supp_2_panel_a<-sup_2_data%>%select(codon, position, interactor, id, replicate, log2FC)%>%
  filter(interactor == "Sho1")%>%
  pivot_wider(names_from = replicate, values_from = log2FC)%>%
  ggpairs(columns = 5:ncol(.), 
          mapping = aes(colour=id), 
          upper = list(continuous = wrap("cor", size = 6, fontface="bold")),
          lower = list(continuous = wrap ("points", alpha=0.25)))+
  ggplot2::scale_color_manual(values = c( "#B52F8CFF", "#F89441FF", "#7E03A8FF"))+
  ggplot2::scale_fill_manual(values = c( "#B52F8CFF", "#F89441FF", "#7E03A8FF"))+
  theme_bw()+
  theme_sup_figure(textsize = 20)+
  theme(title = element_text(size=25, face="bold"),
        strip.text = element_text(size=20))+
  labs(x="Interaction score",
       y="Interaction score",
       title="Sho1")

#Hog1
supp_2_panel_b<-sup_2_data%>%select(codon, position, interactor, id, replicate, log2FC)%>%
  filter(interactor == "Hog1")%>%
  pivot_wider(names_from = replicate, values_from = log2FC)%>%
  ggpairs(columns = 5:ncol(.), 
          mapping = aes(colour=id), 
          upper = list(continuous = wrap("cor", size = 6, fontface="bold")),
          lower = list(continuous = wrap ("points", alpha=0.25)))+
  ggplot2::scale_colour_viridis_d(option="plasma", labels=c("None", "1 M"), begin=0.25, end=0.75)+
  ggplot2::scale_color_manual(values = c( "#B52F8CFF", "#F89441FF", "#7E03A8FF"))+
  ggplot2::scale_fill_manual(values = c( "#B52F8CFF", "#F89441FF", "#7E03A8FF"))+
  theme_bw()+
  theme_sup_figure(textsize=20)+
  theme(title = element_text(size=25, face="bold"),
        strip.text = element_text(size=20))+
  labs(x="Interaction score",
       y="Interaction score",
       title="Hog1")


###ASSEMBLY
supp_fig_2<-plot_grid(ggmatrix_gtable(supp_2_panel_a),
                      ggmatrix_gtable(supp_2_panel_b),
                      nrow=2, ncol=1,
                      labels = "auto",
                      label_size = 30,
                      label_y = c(1.02, 1.02))



supp_fig_2


save_plot(filename="../Data/images/SFigs/Supplementary_figure_2.png", base_asp=0.9 , plot=supp_fig_2, base_height = 15)# , base_asp = 2.0

rm(sup_2_data, supp_2_panel_a, supp_2_panel_b, supp_fig_2)

```



```{r}
####Supp Fig 3 - Sho1 vs Hog1 ####

supp_3_data<-read.csv("../Data/DMS/results/aa_surrounding_region_DMS_scores.csv")%>%
  filter(condition == "MTX" & sorbitol == 1)%>%
  mutate(motif=case_when(position %in% 85:100 ~ "motif",
                         TRUE ~ "non-motif"))




sho1_con_interval<-supp_3_data%>%
  select(position, interactor, aa, type, median_sel_coeff, median_log2FC, sd_log2FC, sd_sel_coeff, motif)%>%
  pivot_wider(names_from = interactor, values_from = c(median_sel_coeff, median_log2FC, sd_log2FC, sd_sel_coeff))%>%
  filter(type == "STOP")%>%
  pull(median_log2FC_Sho1)%>%
  quantile(probs=c(0.025, 0.975), names=F , na.rm=T)



hog1_con_interval<-supp_3_data%>%
  select(position, interactor, aa, type, median_sel_coeff, median_log2FC, sd_log2FC, sd_sel_coeff, motif)%>%
  pivot_wider(names_from = interactor, values_from = c(median_sel_coeff, median_log2FC, sd_log2FC, sd_sel_coeff))%>%
  filter(type == "STOP")%>%
  pull(median_log2FC_Hog1)%>%
  quantile(probs=c(0.025, 0.975), names=F, na.rm=T)



#Grey rectangles represent the 95% confidence interval of the mean of STOP codons
#Values under the 95% quantile of STOP are brought to 97,5 quantile of STOP (detection limit)
supp_fig_3<-supp_3_data %>%
  select(position, interactor, aa, type, median_sel_coeff, median_log2FC, sd_log2FC, sd_sel_coeff, motif)%>%
  pivot_wider(names_from = interactor, values_from = c(median_sel_coeff, median_log2FC, sd_log2FC, sd_sel_coeff))%>%
  filter(type != "STOP")%>%
  mutate(median_log2FC_Sho1 = case_when(median_log2FC_Sho1 < sho1_con_interval[2] ~ sho1_con_interval[2],
                                        TRUE ~ median_log2FC_Sho1),
         median_log2FC_Hog1 = case_when(median_log2FC_Hog1 < hog1_con_interval[2] ~ hog1_con_interval[2],
                                        TRUE ~ median_log2FC_Hog1))%>% #Bring values lower than conf int of STOP codons to limit of con int
  arrange(desc(motif))%>%
ggplot(aes(x=median_log2FC_Sho1, y=median_log2FC_Hog1))+
  geom_rect(xmin=sho1_con_interval[1], xmax=sho1_con_interval[2], ymin=-Inf, ymax=Inf, fill="grey90", colour="grey90")+
  geom_rect(xmin=-Inf, xmax=Inf, ymin=hog1_con_interval[1], ymax=hog1_con_interval[2], fill="grey90", colour="grey90")+
  scale_colour_manual(values = c(motif_colour, flanking_colour), labels=c("Inside Extended Motif", "Outside Extended Motif"))+
  geom_point(aes(colour=motif))+
  geom_abline(intercept=0, slope=1, linetype="dashed")+
  theme_classic()+
  scale_x_continuous(limits =c(-0.7,0.6))+
  scale_y_continuous(limits=c(-0.7, 0.6))+
  theme_sup_figure()+
  theme(legend.title = element_blank(),
       legend.position = c(0.29, 0.85),
       legend.background = element_rect(fill=NA, colour="black"),
       legend.text=element_text(size=9),
       )+
  labs(x="Sho1-Pbs2\nInteraction score",
       y="Hog1-Pbs2\nInteraction score")
  


supp_fig_3_density<-ggMarginal(supp_fig_3, type="density",  groupFill = T) #Add density plots in margins

supp_fig_3_density

save_plot(filename = "../Data/images/SFigs/Supplementary_figure_3.png", plot=supp_fig_3_density)

rm(supp_3_data, sho1_con_interval, hog1_con_interval, supp_fig_3, supp_fig_3_density)
```


```{r, message=F, warning=F}


####Supp Fig 4 - Extended motif replicates####

###Panel B
sup_4_data_motif<-read.csv("../Data/DMS/results/codon_extended_motif_DMS_scores.csv")%>%
  mutate(id=case_when(condition == "MTX" & sorbitol ==1 ~ "MTX + sorbitol",
                      condition == "MTX" & sorbitol ==0 ~ "MTX alone",
                      condition == "DMSO" & sorbitol ==1 ~ "DMSO + sorbitol",
                      condition == "DMSO" & sorbitol ==0 ~ "DMSO alone"))

#Sho1
supp_4_panel_a<-sup_4_data_motif%>%
  arrange(replicate)%>%
  filter(interactor =="Sho1")%>%
  select(codon, position, interactor, replicate, id, log2FC)%>%
  pivot_wider(names_from = replicate, values_from = log2FC)%>%
  ggpairs(columns = 5:ncol(.), 
          mapping = aes(colour=id), 
          upper = list(continuous = wrap("cor", size = 4, fontface="bold")),
          lower = list(continuous = wrap ("points", alpha=0.25)))+
  ggplot2::scale_color_manual(values = c( "#B52F8CFF", "#DE5F65FF", "#F89441FF", "#7E03A8FF"))+
  ggplot2::scale_fill_manual(values = c( "#B52F8CFF", "#DE5F65FF", "#F89441FF", "#7E03A8FF"))+
  theme_bw()+
  theme_sup_figure()+
  theme(title = element_text(size=25, face="bold"),
        strip.text = element_text(size=20))+
  labs(x="Interaction score",
       y="Interaction score",
       title="Sho1")

#Hog1
supp_4_panel_b<-sup_4_data_motif%>%
  arrange(replicate)%>%
  filter(interactor =="Hog1")%>%
  select(codon, position, interactor, replicate, id, log2FC)%>%
  pivot_wider(names_from = replicate, values_from = log2FC)%>%
  ggpairs(columns = 5:ncol(.), 
          mapping = aes(colour=id), 
          upper = list(continuous = wrap("cor", size = 4, fontface="bold")),
          lower = list(continuous = wrap ("points", alpha=0.25)))+
  ggplot2::scale_color_manual(values = c( "#B52F8CFF", "#DE5F65FF", "#F89441FF", "#7E03A8FF"))+
  ggplot2::scale_fill_manual(values = c( "#B52F8CFF", "#DE5F65FF", "#F89441FF", "#7E03A8FF"))+
  theme_bw()+
  theme_sup_figure()+
  theme(title = element_text(size=25, face="bold"),
        strip.text = element_text(size=20))+
  labs(x="Interaction score",
       y="Interaction score",
       title="Hog1")


###ASSEMBLY
supp_fig_4<-plot_grid(ggmatrix_gtable(supp_4_panel_a),
                      ggmatrix_gtable(supp_4_panel_b),
                      nrow=2, ncol=1,
                      labels = "auto",
                      label_size = 30,
                      label_y = c(1.02, 1.02))



supp_fig_4


save_plot(filename="../Data/images/SFigs/Supplementary_figure_4.png", base_asp=0.9 , plot=supp_fig_4, base_height = 15)# , base_asp = 2.0

rm(sup_4_data_motif, supp_4_panel_a, supp_4_panel_b,  supp_fig_4)

```

```{r}
####Supp fig 5 - Corr first/second screen ####

supp_fig5_prelim_data<-read.csv("../Data/DMS/results/aa_surrounding_region_DMS_scores.csv")%>%
  rename(prelim_log2FC=median_log2FC)%>%
  mutate(position=position)
  
supp_fig5_motif_data<-read.csv("../Data/DMS/results/aa_extended_motif_DMS_scores.csv")%>%
  rename(motif_log2FC=median_log2FC)

supp_fig5_data<-left_join(supp_fig5_motif_data, supp_fig5_prelim_data, by=c("position", "aa", "type", "condition", "interactor", "sorbitol", "target"))
  


supp_fig5<-supp_fig5_data%>%
  filter(condition == "MTX" & sorbitol==1 & interactor=="Sho1")%>%
ggplot(aes(x=prelim_log2FC, y=motif_log2FC))+
  geom_point(aes(group=interaction(position, aa)), shape=21, fill=motif_colour)+#
  stat_cor(method="spearman", cor.coef.name = "rho", size=5)+
  theme_classic()+
  theme_sup_figure()+
  labs(x="Sho1-Pbs2 Interaction score\n(Surrounding region)",
       y="Sho1-Pbs2 Interaction score\n(Extended motif)")

supp_fig5

save_plot(filename = "../Data/images/SFigs/Supplementary_figure_5.png", plot = supp_fig5)

rm(supp_fig5_prelim_data, supp_fig5_motif_data, supp_fig5_data, supp_fig5)

```


```{r}
####Supp fig 6 - Destabilization ####

supp_fig6_data<-read.csv("../Data/DMS/results/codon_extended_motif_DMS_scores.csv")%>%
  filter(condition == "MTX" & sorbitol ==1)%>%
          group_by(position, target, condition, sorbitol, interactor, aa, strain_background, type) %>%
          summarize(replicates_per_aa = n(),
                    median_log2FC=median(log2FC, na.rm=T)) %>%
          filter(replicates_per_aa >= 2) %>%
          unite("identifier", target, condition, sorbitol, interactor, strain_background, remove=F)

#T-test to see if Hog1 score for mutants is sig different from wt
supp_fig6_t_test<-read.csv("../Data/DMS/results/codon_extended_motif_DMS_scores.csv")%>%
  filter(condition == "MTX" & interactor %in% c("Sho1", "Hog1") & sorbitol ==1)%>%
  filter(type !="WT" | position == 85)%>%                                        #Only 1 value for WT instead of 1 per position
  dplyr::mutate(identity=case_when(type %in% c("Non-Syn", "STOP") ~ paste(position, aa, sep="_"), #Separate Non-synonymous mutants,
                                   type %in% c("WT", "Syn") ~ "Syn+WT"))%>%          #WT and Syn mutants. the "reference" is WT+Syn pooled
  group_by(condition, interactor, sorbitol)%>%                                       #Group for T-test
  rstatix::pairwise_wilcox_test(log2FC ~ identity,                                   #Non-parametric t-test
                                ref.group = "Syn+WT",                                #Compare Non-syn mutants against Syn+WT   
                                p.adjust.method = "fdr",                             #Correction for multiple corrections
                                alternative ="two.sided")%>%                         #Two sided test                           
  rename(identity = group2, p.signif_log2FC=p.adj.signif, p.adj.log2FC=p.adj)%>%     #For joining later
  select(c(condition, interactor, sorbitol, identity, p.adj.log2FC, p.signif_log2FC)) %>%#Keep only relevant columns
  separate_wider_delim(., cols=identity, delim="_", names=c("position", "aa"))%>%
  mutate(position= as.numeric(position))

supp_fig6_data<-left_join(supp_fig6_data, supp_fig6_t_test, by=c("condition", "interactor", "sorbitol", "position", "aa"))%>%
  ungroup()%>%
  select(position, sorbitol, interactor, aa, type, median_log2FC, p.adj.log2FC, p.signif_log2FC)%>%
   replace_na(list(p.signif_log2FC="ns"))%>%
    pivot_wider(names_from = c(interactor, sorbitol), values_from = c(median_log2FC, p.adj.log2FC, p.signif_log2FC))%>%
filter(!(is.na(median_log2FC_Hog1_1)))


supp_fig6_data<-supp_fig6_data%>%
  mutate(Hog1_effect=case_when(type != "Non-Syn" ~ type,
                               type == "Non-Syn" & p.signif_log2FC_Hog1_1 != "ns" ~ "Non-Syn (Hog1 different)",
                               type == "Non-Syn" & p.signif_log2FC_Hog1_1 == "ns" ~ "Non-Syn (Hog1 not different)"))


supp_fig6<-supp_fig6_data%>%
  arrange(Hog1_effect)%>%
  ggplot(aes(x=median_log2FC_Sho1_1, y=median_log2FC_Hog1_1, colour=Hog1_effect))+
  geom_point(aes(colour=Hog1_effect, group = interaction(aa, position)))+
  scale_colour_manual(values=c("#623CEA", motif_colour, stop_colour, syn_colour, wt_colour), labels=c("Missense (Effect\non Hog1 binding)", "Missense (No effect\non Hog1 binding)", "Nonsense", "Silent", "WT"))+
  theme_classic()+
  theme_sup_figure()+
  theme(legend.text = element_text(margin = margin(t=0.4, b=0.4, unit="line")))+
  labs(x="Sho1-Pbs2 Interaction score",
       y="Hog1-Pbs2 Interaction score",
       colour="Mutation type",
       shape="Hog1 effect")

supp_fig6


 save_plot(filename="../Data/images/SFigs/Supplementary_figure_6.png", plot = supp_fig6, dpi=300, base_asp = 2)

rm(supp_fig6, supp_fig6_data, supp_fig6_t_test)
```

```{r}
####Supp fig 7 - Extended motif heatmap####


#T-test to see if Hog1 score for mutants is sig different from wt
supp_fig7_t_test<-read.csv("../Data/DMS/results/codon_extended_motif_DMS_scores.csv")%>%
  filter(condition == "MTX" & interactor == "Hog1" & sorbitol ==1)%>%
  filter(type !="WT" | position == 85)%>%                                        #Only 1 value for WT instead of 1 per position
  dplyr::mutate(identity=case_when(type %in% c("Non-Syn", "STOP") ~ paste(position, aa, sep="_"), #Separate Non-synonymous mutants,
                                   type %in% c("WT", "Syn") ~ "Syn+WT"))%>%          #WT and Syn mutants. the "reference" is WT+Syn pooled
  group_by(condition, interactor, sorbitol)%>%                                       #Group for T-test
  rstatix::pairwise_wilcox_test(log2FC ~ identity,                                   #Non-parametric t-test
                                ref.group = "Syn+WT",                                #Compare Non-syn mutants against Syn+WT   
                                p.adjust.method = "fdr",                             #Correction for multiple corrections
                                alternative ="two.sided")%>%                         #Two sided test                           
  rename(identity = group2)%>%                                                       #For joining later
  select(c(condition, interactor, sorbitol, identity, p.adj, p.adj.signif)) %>%      #Keep only relevant columns
  separate_wider_delim(., cols=identity, delim="_", names=c("position", "aa"))%>%
  mutate(position= as.numeric(position))%>%
  filter(p.adj.signif!="ns" & aa != "*")%>%
  select(position, aa)


supp_7_data<-read.csv("../Data/DMS/results/codon_extended_motif_DMS_scores.csv")%>%
  filter(condition == "MTX" & sorbitol ==1 & interactor== "Sho1")%>%
          group_by(position, target, condition, sorbitol, interactor, aa, strain_background, type) %>%
          summarize(replicates_per_aa = n(),
                    median_log2FC=median(log2FC, na.rm=T)) %>%
          filter(replicates_per_aa >= 2) %>%
          unite("identifier", target, condition, sorbitol, interactor, strain_background, remove=F)
  

supp_fig7_wt_pbs2<-read.csv("../Data/DMS/reference/Pbs2_wt_seq.csv")%>%
  mutate(position=position+70)%>%
  filter(position %in% 85:100)

supp_fig_7<-supp_7_data%>%
  ggplot(aes(x=as.factor(position), y=aa))+
  geom_tile(aes(fill=median_log2FC, colour=median_log2FC))+
  geom_text(data=supp_fig7_wt_pbs2, label="*")+
  geom_text(data=supp_fig7_t_test, label="H")+
  # geom_tile(data=supp_fig6_t_test, fill=NA, colour="black", linewidth=1)+
  scale_x_discrete(expand=c(0,0), )+
  scale_y_discrete(expand = c(0,0))+
  scale_fill_gradient2(low="blue", mid="white", high="red")+
  scale_colour_gradient2(low="blue", mid="white", high="red", guide="none")+
  theme_sup_figure()+
  theme(panel.border = element_rect(fill = NA, colour="black"),
        panel.background = element_rect(fill="black"),
          panel.grid.major = element_blank(),
        axis.text.x=element_text(size=15))+
    labs(x="Position in Pbs2",
         y="Residue",
         fill="Sho1-Pbs2\nInteraction score")


supp_fig_7

save_plot(filename = "../Data/images/SFigs/Supplementary_figure_7.png", base_width=8 ,plot=supp_fig_7, dpi=300)

rm(supp_7_data, supp_fig7_wt_pbs2, supp_fig_7, supp_fig7_t_test)

```


```{r}

####Supp fig  8 - Effect of sorbitol####

supp_fig8_data<-read.csv("../Data/DMs/results/aa_extended_motif_DMS_scores.csv")%>%
  filter(interactor =="Sho1" & condition == "MTX")%>%
  select(position, aa, sorbitol, type, median_log2FC)%>%
  pivot_wider(names_from = sorbitol, values_from = median_log2FC, names_prefix = "sorb_")


supp_fig8<-supp_fig8_data%>%
  ggplot(aes(x=sorb_1, y=sorb_0))+
  geom_point(aes(colour=type))+
  stat_cor(method="spearman", cor.coef.name = "rho", show.legend = F, size=6)+
  scale_colour_manual(values=c(motif_colour, stop_colour, syn_colour, wt_colour),labels=c("Missense", "Nonsense", "Silent", "WT"))+
  theme_classic()+
  theme_sup_figure()+
  labs(x="Sho1-Pbs2 Interaction score\nin 1 M sorbitol",
       y="Sho1-Pbs2 Interaction score\nin absence of sorbitol",
       colour="Mutation type")

supp_fig8

save_plot(filename="../Data/images/SFigs/Supplementary_figure_8.png", plot=supp_fig8, dpi=300)

rm(supp_fig8, supp_fig8_data)
```


```{r}

####Supp fig 9 - Validation####



# DMS data from the first DMS screen, larger region
supp_fig9_prelim_dms_data<-read.csv("../Data/DMS/results/aa_surrounding_region_DMS_scores.csv")%>%
  filter(condition == "MTX" &  interactor == "Sho1")%>%
  select(median_log2FC, sd_log2FC, position, aa, sorbitol)%>%
  mutate(screen="Surrounding region")

#DMS data from second DMS screen on extended motif
supp_fig9_motif_dms_data<-read.csv("../Data/DMS/results/aa_extended_motif_DMS_scores.csv")%>%
  filter(condition == "MTX" &  interactor == "Sho1")%>%
  select(median_log2FC, sd_log2FC, position, aa, sorbitol)%>%
  mutate(screen="Extended motif")

#combine DMS data
supp_fig9_dms_data<-rbind(supp_fig9_prelim_dms_data, supp_fig9_motif_dms_data)

#PANEL A - growth curves
supp_fig9a_gc_data<-read.csv("../Data/growthcurves/results_validation_growth_curves.csv")

#Prepare growth curve data frame for joining with DMS data
supp_fig9a_gc_data<-supp_fig9a_gc_data%>%
  filter(mutation != "P94A+P97A")%>% #Not in DMS data, as it is a double mutant
  group_by(mutation, condition)%>%
  summarize(median_mtx_gr=median(mtx_gr, na.rm=T),
            sd_mtx_gr=sd(mtx_gr, na.rm=T),
            median_dmso_gr=median(dmso_gr, na.rm=T),
            sd_dmso_gr=sd(dmso_gr, na.rm=T),
            replicates=n())%>%
  ungroup%>%
  mutate(sorbitol=case_when(condition == "Sorbitol" ~ 1,
                            TRUE ~ 0))%>%
mutate(position=as.numeric(str_replace_all(mutation, "[:alpha:]", "")),
       aa=str_sub(mutation, start=-1))

#Combine DMS data and growth curve data
supp_fig9a_data<-left_join(supp_fig9a_gc_data, supp_fig9_dms_data, by=c("position", "aa", "sorbitol"))%>%
  drop_na("screen")


#Validation of fitness effects.
supp_fig9_panel_a<-supp_fig9a_data%>%
  mutate(sorbitol=as.character(sorbitol))%>%
  ggplot(aes(x=median_log2FC, y=median_mtx_gr, colour=sorbitol))+
    geom_point(aes(group=mutation))+
    scale_colour_viridis_d(option="plasma", labels=c("None", "1 M"), begin=0.25, end=0.75)+
    stat_cor(method="spearman", cor.coef.name = "rho", show.legend=F, size=5)+
  facet_wrap(~screen)+
    theme_classic()+
  theme_sup_figure(textsize=17)+
  theme(panel.background = element_rect(colour="black"),
        strip.text = element_text(size=20),
        legend.key = element_rect(colour = "white"), 
        panel.spacing.x = unit(1.2, "lines"))+
    labs(x="DMS libraries Sho1-Pbs2 Interaction scores",
       y="Reconstructed mutants\nGrowth rate (OD/h)",
       colour="Sorbitol")

supp_fig9_panel_a

#Panel B - validation DHFR-PCA

supp_fig9b_validation<-read.csv("../Data/DMS/results/aa_validation_scores.csv")

supp_fig9b_validation<-supp_fig9b_validation%>%
  filter(condition == "MTX" & interactor == "Sho1")%>%
  select(median_log2FC, position, aa, sorbitol)%>%
  rename(validation_median_log2FC=median_log2FC)

supp_fig9b_data<-inner_join(supp_fig9b_validation, supp_fig9_dms_data, by=c("aa", "position", "sorbitol"))


supp_fig9_panel_b<-supp_fig9b_data%>%
  mutate(sorbitol=as.character(sorbitol))%>%
  ggplot(aes(x=median_log2FC, y=validation_median_log2FC, colour=sorbitol))+
    geom_point(aes(group=interaction(aa, position)))+
    scale_colour_viridis_d(option="plasma", labels=c("None", "1 M"), begin=0.25, end=0.75)+
    stat_cor(method="spearman", cor.coef.name = "rho", show.legend=F, size=5)+
  facet_wrap(~screen)+
    theme_classic()+
  theme_sup_figure(textsize=17)+
  theme(panel.background = element_rect(colour="black"),
        strip.text = element_text(size=20),
        legend.key = element_rect(colour = "white"),
        panel.spacing.x = unit(1.2, "lines"))+
    labs(x="DMS libraries Sho1-Pbs2 Interaction scores",
       y="Reconstructed mutants\nSho1-Pbs2 Interaction score",
       colour="Sorbitol")

supp_fig9_panel_b


supp_fig9<-cowplot::plot_grid(supp_fig9_panel_a+theme(legend.text = element_text(size=17),
                                                      legend.title = element_text(size=20)), 
                              supp_fig9_panel_b+theme(legend.text = element_text(size=17),
                                                      legend.title = element_text(size=20)), 
                              nrow=2,ncol=1,
                              labels = "auto", label_colour = "black", label_size = 20)

supp_fig9

save_plot(filename = "../Data/images/SFigs/Supplementary_figure_9.png", supp_fig9, dpi=300, base_height = 8)


rm(supp_fig9, supp_fig9_dms_data, supp_fig9_motif_dms_data, supp_fig9_panel_a, supp_fig9_panel_b, supp_fig9_prelim_dms_data, supp_fig9a_data, supp_fig9a_gc_data, supp_fig9b_data, supp_fig9b_validation)
```

```{r, message=F, warning=F}
#### Figure S10 - Validation replicates####


###Panel C
sup_10_data_validation<-read.csv("../Data/DMS/results/codon_validation_scores.csv")%>%
  mutate(interactor=case_when(interactor == "None" ~ "Proliferation",
                              TRUE ~ interactor))%>%
  mutate(id=case_when(condition == "MTX" & sorbitol ==1 ~ "MTX + sorbitol",
                      condition == "MTX" & sorbitol ==0 ~ "MTX alone",
                      condition == "DMSO" & sorbitol ==1 ~ "DMSO + sorbitol",
                      condition == "DMSO" & sorbitol ==0 ~ "DMSO alone",
                      condition == "Stress" & sorbitol == 1 ~ "Sorbitol",
                      condition == "Stress" & sorbitol == 0 ~ "No sorbitol"))

#Sho1
supp_10_panel_a<-sup_10_data_validation%>%
  arrange(replicate)%>%
  filter(interactor =="Sho1")%>%
  select(codon, position, interactor, id, replicate, log2FC)%>%
  pivot_wider(names_from = replicate, values_from = log2FC)%>%
  ggpairs(columns = 5:ncol(.), 
          mapping = aes(colour=id), 
          upper = list(continuous = wrap("cor", size = 4, fontface="bold")),
          lower = list(continuous = wrap ("points", alpha=0.25)))+
  ggplot2::scale_color_manual(values = c( "#B52F8CFF", "#DE5F65FF", "#F89441FF", "#7E03A8FF"))+
  ggplot2::scale_fill_manual(values = c( "#B52F8CFF", "#DE5F65FF", "#F89441FF", "#7E03A8FF"))+
  theme_bw()+
  theme_sup_figure()+
  theme(title = element_text(size=25, face="bold"),
        strip.text = element_text(size=20),
        axis.text.x = element_text(size=15))+
  labs(x="Interaction score",
       y="Interaction score", 
       title="Sho1")

#Hog1
supp_10_panel_b<-sup_10_data_validation%>%
  arrange(replicate)%>%
  filter(interactor =="Hog1")%>%
  select(codon, position, interactor, id, replicate, log2FC)%>%
  pivot_wider(names_from = replicate, values_from = log2FC)%>%
  ggpairs(columns = 5:ncol(.), 
          mapping = aes(colour=id), 
          upper = list(continuous = wrap("cor", size = 4, fontface="bold")),
          lower = list(continuous = wrap ("points", alpha=0.25)))+
  ggplot2::scale_color_manual(values = c( "#B52F8CFF", "#DE5F65FF", "#F89441FF", "#7E03A8FF"))+
  ggplot2::scale_fill_manual(values = c( "#B52F8CFF", "#DE5F65FF", "#F89441FF", "#7E03A8FF"))+
  theme_bw()+
  theme_sup_figure()+
  theme(title = element_text(size=25, face="bold"),
        strip.text = element_text(size=20),
        axis.text.x = element_text(size=15))+
  labs(x="Interaction score",
       y="Interaction score", 
       title="Hog1")

#Proliferation
supp_10_panel_c<-sup_10_data_validation%>%
  arrange(replicate)%>%
  filter(interactor =="Proliferation")%>%
  select(codon, position, interactor, id, replicate, selection_coefficient)%>%
  pivot_wider(names_from = replicate, values_from = selection_coefficient)%>%
  ggpairs(columns = 5:ncol(.), 
          mapping = aes(colour=id), 
          upper = list(continuous = wrap("cor", size = 5, fontface="bold")),
          lower = list(continuous = wrap ("points", alpha=0.25)))+
  ggplot2::scale_color_manual(values = c("#7E03A8FF", "#F89441FF"))+
  ggplot2::scale_fill_manual(values = c("#7E03A8FF", "#F89441FF"))+
  theme_bw()+
  theme_sup_figure()+
  theme(title = element_text(size=25, face="bold"),
        strip.text = element_text(size=20),
        axis.text.x = element_text(size=15))+
  labs(x="Selection coefficient",
       y="Selection coefficient", 
       title="Proliferation")


###ASSEMBLY
supp_fig_10<-plot_grid(ggmatrix_gtable(supp_10_panel_a),
                      ggmatrix_gtable(supp_10_panel_b),
                      ggmatrix_gtable(supp_10_panel_c),
                      nrow=3, ncol=1,
                      labels = "auto",
                      label_size = 30,
                      label_y = c(1.02, 1.02))



supp_fig_10


save_plot(filename="../Data/images/SFigs/Supplementary_figure_10.png", base_asp=0.9 , plot=supp_fig_10, base_height = 20)# , base_asp = 2.0

rm(sup_10_data_validation, supp_10_panel_a, supp_10_panel_b, supp_10_panel_c, supp_fig_10)
```

```{r}

####Supp fig 11 - Fitness effects ####


supp_fig11_data<-read.csv("../Data/DMS/results/codon_validation_scores.csv")%>%
  filter(condition == "Stress")


supp_fig11_t_test<-supp_fig11_data%>%
  filter(condition == "Stress")%>%
  dplyr::mutate(t_test_type=case_when(type %in% c("Non-Syn", "STOP") ~ mutant, #Separate Non-syn mutants,
                                   type %in% c("WT", "Syn") ~ "Syn+WT"))%>%          #the "reference" is WT+Syn pooled
  group_by(sorbitol)%>%                                       #Group for T-test
  rstatix::pairwise_t_test(selection_coefficient ~ t_test_type,                                   #Non-parametric t-test
                                ref.group = "Syn+WT",                                #Compare Non-syn mutants against Syn+WT   
                                p.adjust.method = "fdr",                             #Correction for multiple corrections
                                alternative ="two.sided",                            #Two-sided t-test 
                           pool.sd=F,                                               #Welch's t-test
                           var.equal=F)%>%                         
  rename(mutant = group2, p.signif_sel_coeff=p.adj.signif, p.adj.sel_coeff=p.adj)


supp_fig11<-supp_fig11_data%>%
  filter(sorbitol ==1)%>%
  ggplot(aes(x=mutant, y=selection_coefficient))+
  geom_text(data=supp_fig11_t_test%>%filter(p.signif_sel_coeff != "ns" & sorbitol ==1), aes(x=mutant, y=0.1, label=p.signif_sel_coeff))+
  geom_boxplot(aes(fill=type), outlier.shape = 21, outlier.size = 0.75)+
  scale_fill_manual(values = c(motif_colour, stop_colour, syn_colour, wt_colour), labels=c("Missense", "Nonsense", "Silent", "WT"), name="Mutation type")+
  theme_bw()+
  theme_sup_figure(textsize = 15)+
  theme(axis.text.x = element_text(angle=90, hjust=1, vjust=0.5, 
                                   colour=c("black", "black", "red", rep("black", 58)),
                                   face=c("plain","plain", "bold", rep("plain", 58)),
                                   size=13))+
  labs(x="Pbs2 mutant",
       y="Selection coefficient")


supp_fig11

save_plot(filename = "../Data/images/SFigs/Supplementary_figure_11.png", plot=supp_fig11, base_asp = 3.5)


rm(supp_fig11_data, supp_fig11, supp_fig11_t_test)

```


```{r}
####Supp figure 12 - AF2 metrics ####

#PANEL A

supp_12a_data<-read.csv("../Data/AlphaFold_predictions/alphafold_output/PAE.csv")

supp_fig_12a<-supp_12a_data%>%
  ggplot(aes(x=scored_residue, y=aligned_residue, fill=PAE))+
  geom_tile()+
    scale_y_reverse(expand=c(0,0))+
    scale_x_continuous(expand=c(0,0))+
    scale_fill_gradient2(low="#022811", mid="#4BAF4D", high="#F8FDF6", breaks=c(0,5,10,15,20,25,30))+
    theme_sup_figure()+
    theme(
      # aspect.ratio = 1,
        panel.border = element_rect(colour="black", fill = NA),
        title = element_text(size=8),
        axis.text = element_text(colour="black"),
        strip.background=element_rect(fill=NA, colour=NA),
          strip.text = element_text(face="bold", size=10))+
  facet_wrap(~model, scales="free")+
    guides(fill=guide_colourbar(title.position = "top", 
                              frame.colour = "black", 
                              ticks.colour = "black", 
                              title.hjust =0.5))+
    labs(x="Scored residue",
         y="Aligned residue",
         fill="Expected position\nerror (Ångström)")

supp_fig_12a


#PANEL B

supp_12b_data<-read.csv("../Data/AlphaFold_predictions/alphafold_output/pLDDT.csv")

models<-supp_12b_data%>%pull(model)%>%unique()
supp_12_sh3_domain<-data.frame(begin=c(302,0,302,302,302,302,302),
                      end=c(360,58,360,360,360,360,360),
                      model=models)
supp_12_extended_motif<-data.frame(begin=c(381,72,381,381,381,381,381),
                           end=c(397,88,397,397,397,397,397),
                           model=models)
supp_12_limit<-data.frame(value=c(367, 58,367,367,367,367,367),
                  model=models)

supp_fig_12b<-supp_12b_data%>%
  ggplot()+
    geom_rect(data=supp_12_extended_motif, ymin=-Inf, ymax=Inf, aes(xmin=begin, xmax=end),fill="skyblue")+
    geom_rect(data=supp_12_sh3_domain, aes(xmin=begin, xmax=end), ymin=-Inf, ymax=Inf, fill="#028b5b")+
    geom_line(aes(x=position, y = plddt))+
    geom_vline(data=supp_12_limit, aes(xintercept=value), linetype="dashed", colour="red", linewidth=0.75)+
    scale_x_continuous(expand=c(0,0))+
    facet_wrap(~model, scales="free_x")+
    theme_classic()+
    theme_sup_figure(textsize = 15)+
    theme(strip.background=element_rect(fill=NA),
          strip.text = element_text(face="bold", size=10),)+
    labs(y="pLDDT", 
         x="Position in model")

supp_fig_12b


#ASSEMBLY

supp_fig_12<-plot_grid(supp_fig_12a,
                       NULL,
                       supp_fig_12b,
                       nrow=3, ncol=1,
                       rel_heights = c(1,0.05,1),
                       labels=c("a","", "b"), label_size = 20,
                       align="v",
                       axis="l")+theme(panel.background = element_rect("white", colour="white"))

supp_fig_12

save_plot(filename="../Data/images/SFigs/Supplementary_figure_12.png", plot=supp_fig_12, base_height=10, base_asp = 0.75, dpi=300)

rm(supp_12_extended_motif, supp_12_limit, supp_12_sh3_domain, supp_12a_data, supp_12b_data, supp_fig_12, supp_fig_12a, supp_fig_12b)

```


```{r}
####Supp fig 13 - Structures####

supp_fig13a_H86W<-"../Data/images/SFigs/S13/H86W_review.png"
supp_fig13b_V91L<-"../Data/images/SFigs/S13/V91L_review.png"
supp_fig13c_V91M<-"../Data/images/SFigs/S13/V91M_review.png"
supp_fig13d_Q89D<-"../Data/images/SFigs/S13/Q89D_review.png"

supp_fig13a_panel_H86W<-ggdraw()+draw_image(supp_fig13a_H86W, scale=1)
supp_fig13b_panel_V91L<-ggdraw()+draw_image(supp_fig13b_V91L, scale=1)
supp_fig13c_panel_V91M<-ggdraw()+draw_image(supp_fig13c_V91M, scale=1)
supp_fig13d_panel_Q89D<-ggdraw()+draw_image(supp_fig13d_Q89D, scale=1)


supp_fig13<-plot_grid(supp_fig13a_panel_H86W,
                      supp_fig13b_panel_V91L,
                      supp_fig13c_panel_V91M,
                      supp_fig13d_panel_Q89D,
                     labels = "auto", label_colour="black", label_x = 0, label_y = 1.05, label_size = 12,
                     nrow = 2, ncol = 2)+
  theme(panel.background=element_rect(fill="white", colour="white"))

supp_fig13


save_plot("../Data/images/SFigs/Supplementary_figure_13.png", supp_fig13, ncol=2, nrow=2, dpi=500, base_height=1)

rm(supp_fig13, supp_fig13a_panel_H86W, supp_fig13b_panel_V91L, supp_fig13c_panel_V91M, supp_fig13d_panel_Q89D, supp_fig13a_H86W, supp_fig13b_V91L, supp_fig13c_V91M, supp_fig13d_Q89D)

```

```{r}
####Supplementary Figure 14 - Sequence and structure alignment####

#Input files
supp_figure_14_sequence_alignment<- "../Data/SH3_alignments/mafft_SH3_aligned.fasta"
supp_figure_14_structure_alignment<-"../Data/SH3_alignments/MUSTANG_structural_alignment.afasta"

#Make alignment figures one line at a time

supp_fig14_panel_a_top<-ggmsa(supp_figure_14_sequence_alignment, start=1, end=71, seq_name=T, border = NA, char_width=0.5)+
  theme(axis.text.x = element_text(colour="black"),
        axis.text.y = element_text(colour="black", size=6))
supp_fig14_panel_a_bottom<-ggmsa(supp_figure_14_sequence_alignment, start=72, end=145, seq_name=T, border=NA, char_width=0.5)+
  theme(axis.text.x = element_text(colour="black"),
        axis.text.y = element_text(colour="black", size=6))


supp_fig14_panel_b_top<-ggmsa(supp_figure_14_structure_alignment, start = 1, end=81, seq_name = T, border=NA, char_width = 0.5)+
  theme(axis.text.x = element_text(colour="black"),
        axis.text.y = element_text(colour="black", size=6))
  supp_fig14_panel_b_middle<-ggmsa(supp_figure_14_structure_alignment, start = 82, end=162, seq_name = T, border=NA, char_width=0.5)+
  theme(axis.text.x = element_text(colour="black"),
        axis.text.y = element_text(colour="black", size=6))
supp_fig14_panel_b_bottom<-ggmsa(supp_figure_14_structure_alignment, start = 163, seq_name = T, border=NA, char_width=0.5)+
  theme(axis.text.x = element_text(colour="black"),
        axis.text.y = element_text(colour="black", size=6))


#Assemble a and b separately
supp_fig14_panel_a<-plot_grid(supp_fig14_panel_a_top,
                              supp_fig14_panel_a_bottom,
                              ncol=1, nrow=2)

supp_fig14_panel_b<-plot_grid(supp_fig14_panel_b_top,
          supp_fig14_panel_b_middle,
          supp_fig14_panel_b_bottom,
          ncol=1, nrow=3)

supp_fig14_panel_a
supp_fig14_panel_b



save_plot(filename="../Data/images/SFigs/S14/Supplementary_Figure_14_a.png", plot=supp_fig14_panel_a, base_height = 5.57, base_width = 8, bg="white")
save_plot(filename="../Data/images/SFigs/S14/Supplementary_Figure_14_b.png", plot=supp_fig14_panel_b, base_height = 7.42, base_width = 8, bg="white")




### Modify the subplots above using Inkscape

supp_fig14a_mod<-"../Data/images/SFigs/S14/S14_a_mod.png"
supp_fig14b_mod<-"../Data/images/SFigs/S14/S14_b_mod.png"


supp_fig14a_mod_draw<-ggdraw()+draw_image(supp_fig14a_mod, scale=1)
supp_fig14b_mod_draw<-ggdraw()+draw_image(supp_fig14b_mod, scale=1)

supp_fig14_assembled<-plot_grid(supp_fig14a_mod_draw,
          supp_fig14b_mod_draw,
          nrow=2,
          labels="auto", rel_widths = c(0.95, 1), hjust=-3)

save_plot(filename= "../Data/images/SFigs/Supplementary_Figure_14.png", plot=supp_fig14_assembled, ncol=1, base_width = 8, base_height = 13, bg="white")

rm(list=ls(pattern="supp_fig14"))


```

```{r}
####Supplementary Figure 15 - Pbs2 interaction with all SH3####


supp_15_data<-read.csv("../Data/Solid_PCA/results/solid_pca_results.csv")%>%
  mutate(partner=str_to_title(partner))

supp_fig_15<-supp_15_data%>%
  ggplot(aes(x=partner, y=mednorm, fill=mutation))+
  geom_boxplot(outlier.shape=NA)+
  geom_point(position=position_jitterdodge(jitter.width=0.25), shape=21, show.legend = F)+
  scale_fill_discrete(labels=c("Stuffed", "WT"), type=c("white", wt_colour))+
  facet_wrap(facets = vars(condition), nrow=2, ncol=1)+
  theme_bw()+
  theme_sup_figure(textsize=15)+
  theme(axis.text.x = element_text(angle=45, hjust=1, vjust=1, size=13),
        strip.text = element_text(size=15))+
  labs(fill="Pbs2 variant",
       y="Normalized interaction score",
       x=NULL)
  
supp_fig_15

save_plot(filename = "../Data/images/SFigs/Supplementary_Figure_15.png", plot = supp_fig_15, base_width= 8, base_height = 5 )

rm(supp_15_data, supp_fig_15)



```

```{r}
####Supplementary Figure 16 - Superimposed structures####

#Read images
abp1_pbs2<-"../Data/images/SFigs/S16/sho1-abp1.png"
nbp2_pbs2<-"../Data/images/SFigs/S16/sho1-nbp2.png"
bem1_pbs2<-"../Data/images/SFigs/S16/sho1-bem1.png"

#Draw images as gg objects
pic_abp1_pbs2<-ggdraw()+draw_image(abp1_pbs2, scale=1)
pic_nbp2_pbs2<-ggdraw()+draw_image(nbp2_pbs2, scale=1)
pic_bem1_pbs2<-ggdraw()+draw_image(bem1_pbs2, scale=1)

#Assemble and annotate images
supp_fig_16<-plot_grid(pic_abp1_pbs2,
                       pic_nbp2_pbs2,
                       pic_bem1_pbs2,
                       labels = "auto", label_size = 20)+
  annotate(geom="label", x=c(0.1,0.6,0.1), y=c(0.6,0.6,0.1), label=c("Sho1-Abp1", "Sho1-Nbp2", "Sho1-Bem1"), colour="black", size=5)



save_plot(filename = "../Data/images/SFigs/Supplementary_Figure_16.png", plot = supp_fig_16, bg="white")

rm(abp1_pbs2, nbp2_pbs2, bem1_pbs2, pic_abp1_pbs2, pic_bem1_pbs2, pic_nbp2_pbs2, supp_fig_16)

```



```{r}

####Supplementary Figure 17 -Additional differences in binding####




#Panel A - position -7
#Read images
pbs2_7<-"../Data/images/SFigs/S17/annotations_Pbs2_7.png"
ark1_7_contacts<-"../Data/images/SFigs/S17/annotations_Ark1_7.png"
ste20_nbp2_7_contacts<-"../Data/images/SFigs/S17/annotations_nbp2_7.png"
ste20_bem1_7_contacts<-"../Data/images/SFigs/S17/annotations_bem1_7.png"


#Draw images as gg objects
pic_pbs2_7<-ggdraw()+draw_image(pbs2_7, scale=1)
pic_ark1_7_contacts<-ggdraw()+draw_image(ark1_7_contacts, scale=1)
pic_ste20_nbp2_7_contacts<-ggdraw()+draw_image(ste20_nbp2_7_contacts, scale=1)
pic_ste20_bem1_7_contacts<-ggdraw()+draw_image(ste20_bem1_7_contacts, scale=1)

#Assemble and annotate images
position_minus_7<-plot_grid(pic_pbs2_7,
          pic_ark1_7_contacts,
          pic_ste20_nbp2_7_contacts,
          pic_ste20_bem1_7_contacts)+
  annotate(geom="label", x=c(0.1,0.6,0.1,0.6), y=c(0.9,0.9,0.4,0.075), label=c("Sho1-Pbs2", "Abp1-Ark1", "Nbp2-Ste20", "Bem1-Ste20"), colour="black", size=5)



# Panel B - position -4

#Read images
pbs2_4<-"../Data/images/SFigs/S17/annotations_Pbs2_4.png"
ark1_4_contacts<-"../Data/images/SFigs/S17/annotations_Ark1_4.png"
ste20_nbp2_4_contacts<-"../Data/images/SFigs/S17/annotations_nbp2_4.png"
ste20_bem1_4_contacts<-"../Data/images/SFigs/S17/annotations_bem1_4.png"



#Draw images as gg objects
pic_pbs2_4<-ggdraw()+draw_image(pbs2_4, scale=1)
pic_ark1_4_contacts<-ggdraw()+draw_image(ark1_4_contacts, scale=1)
pic_ste20_nbp2_4_contacts<-ggdraw()+draw_image(ste20_nbp2_4_contacts, scale=1)
pic_ste20_bem1_4_contacts<-ggdraw()+draw_image(ste20_bem1_4_contacts, scale=1)

#Assemble and annotate images
position_minus_4<-plot_grid(pic_pbs2_4,
          pic_ark1_4_contacts,
          pic_ste20_nbp2_4_contacts,
          pic_ste20_bem1_4_contacts)+
  annotate(geom="label", x=c(0.1,0.6,0.1,0.6), y=c(0.9,0.9,0.4,0.1), label=c("Sho1-Pbs2", "Abp1-Ark1", "Nbp2-Ste20", "Bem1-Ste20"), colour="black", size=5)


# Assemble both panels into one
supp_fig_17<-plot_grid(position_minus_7,
                 ggplot()+theme_classic(), #extra spacing between panels
                 position_minus_4,
                 labels=c("a", "", "b"), label_size = 30, label_y=1.01, 
                 nrow=3, ncol=1,
                 rel_heights = c(1,0.05,1))


cowplot::save_plot(filename = "../Data/images/SFigs/Supplementary_Figure_17.png", plot = supp_fig_17, ncol=1, nrow=3, bg="white", base_asp = 2.2)




rm(pic_ste20_bem1_4_contacts, pic_ste20_bem1_7_contacts, pic_ste20_nbp2_4_contacts, pic_ste20_nbp2_7_contacts, ste20_bem1_4_contacts, ste20_bem1_7_contacts, ste20_nbp2_4_contacts, ste20_nbp2_7_contacts, pbs2_4, pbs2_7, pic_pbs2_4, pic_pbs2_7, ark1_4_contacts, ark1_7_contacts, pic_ark1_4_contacts, pic_ark1_7_contacts, position_minus_4, position_minus_7, supp_fig_17)

```

